includes/Classifai/Providers/Azure/Personalizer.php

```diff
@@ -0,0 +1,604 @@
+						__( 'Azure Cognitive Service Personalizer Endpoint, <a href="%1$s" target="_blank">create a Personalizer resource</a> in the Azure portal to get your key and endpoint.', 'classifai' ),
+						array(
+							'a' => array(
+								'href'   => array(),
+								'target' => array(),
+							),
+						)
+					),
+					esc_url( 'https://portal.azure.com/#create/Microsoft.CognitiveServicesPersonalizer' )
+				),
+			]
+		);
+		add_settings_field(
+			'api-key',
+			esc_html__( 'API Key', 'classifai' ),
+			[ $this, 'render_input' ],
+			$this->get_option_name(),
+			$this->get_option_name(),
+			[
+				'label_for'     => 'api_key',
+				'input_type'    => 'password',
+				'default_value' => $default_settings['api_key'],
+				'description'   => __( 'Azure Cognitive Service Personalizer Key.', 'classifai' ),
+			]
+		);
+	}
+
+	/**
+	 * Sanitize settings.
+	 *
+	 * @param array $settings The settings being saved.
+	 *
+	 * @return array|mixed
+	 */
+	public function sanitize_settings( $settings ) {
+		$new_settings = [];
+		if ( ! empty( $settings['url'] ) && ! empty( $settings['api_key'] ) ) {
+			$auth_check = $this->authenticate_credentials( $settings['url'], $settings['api_key'] );
+			if ( is_wp_error( $auth_check ) ) {
+				$settings_errors['classifai-registration-credentials-error'] = $auth_check->get_error_message();
+				$new_settings['authenticated']                               = false;
+			} else {
+				$new_settings['authenticated'] = true;
+			}
+			$new_settings['url']     = esc_url_raw( $settings['url'] );
+			$new_settings['api_key'] = sanitize_text_field( $settings['api_key'] );
+		} else {
+			$new_settings['authenticated'] = false;
+			$new_settings['url']           = '';
+			$new_settings['api_key']       = '';
+
+			$settings_errors['classifai-registration-credentials-empty'] = __( 'Please enter your credentials', 'classifai' );
+		}
+
+		if ( ! empty( $settings_errors ) ) {
+
+			$registered_settings_errors = wp_list_pluck( get_settings_errors( $this->get_option_name() ), 'code' );
+
+			foreach ( $settings_errors as $code => $message ) {
+
+				if ( ! in_array( $code, $registered_settings_errors, true ) ) {
+					add_settings_error(
+						$this->get_option_name(),
+						$code,
+						esc_html( $message ),
+						'error'
+					);
+				}
+			}
+		}
+
+		return $new_settings;
+	}
+
+	/**
+	 * Get Recent posts based on given arguments.
+	 *
+	 * @param array $attributes The block attributes.
+	 * @return array recent actions based on block attributes.
+	 */
+	protected function get_recent_actions( $attributes ) {
+		$post_type      = $attributes['contentPostType'];
+		$key_attributes = array(
+			'terms' => isset( $attributes['taxQuery'] ) ? $attributes['taxQuery'] : array(),
+		);
+		$transient_key  = 'classifai_actions_' . $post_type . md5( maybe_serialize( $key_attributes ) );
+		$actions        = get_transient( $transient_key );
+		if ( false === $actions ) {
+			$query_args = array(
```

Comment:
I'm not seeing this handled anywhere else but we will want to ensure that we don't ever recommend the current item that is being displayed. Not sure if that means we exclude that here (`post__not_in`?) or remove it later but I think with the way this is set up right now, you could end up showing a recommended piece of content on that same piece of content itself.